File: src/fenceai/core/gaps.py
"""`Gap` — what is missing, said addressably (integration contract §1.2.1).

One type whether the Knowledge Platform publishes it or a planning run produces
it. That is the contract's wording and it is the reason this lives in `core/`
rather than under `knowledge/` or `strategy/`: a gap arriving in a snapshot and a
gap discovered while laying out a fence are the same fact seen from two sides,
and a curator's queue has to be able to hold both without a translation step.

**A gap is not a failure and not a warning.** A failure means the run stops; a
warning is a note on an answer. A gap is a hole in what we were told, carried
alongside an answer that was produced anyway — contract §3.2.4: *never fail a run
over a gap; warned, named, unfulfilled lines instead.* Every gap here is paired
with a `StrategyWarning` so the hole is visible on the drawing, and with a `gap`
node in the decision graph so it is traceable to the line it affected.

The fields are the contract's, unrenamed. Two are BINDING and are the reason the
type is worth having at all:

* `would_close` — one sentence naming what would resolve it. *"a footing row for
  exposure C, non-HVHZ, at 6 ft"* is a work item; *"footing depth is missing"*
  sends a curator hunting.
* `closes_by` — WHO can close it. Two kinds (`unmodellable_entity`,
  `unmapped_part_kind`) close by a schema change in THIS repo and by nothing a
  curator can do, and a review queue that shows a curator work only an engineer
  can perform is a queue whose items are not actionable.
"""

from __future__ import annotations

import json
from typing import Literal

from pydantic import BaseModel, model_validator

GapKind = Literal[
    "unmodellable_entity",       # the corpus describes something no type fits
    "uncovered_condition",       # a domain point no row covers
    "unsatisfiable_requirement", # nothing can fill a slot
    "unquantified",              # stated in prose, never given a number
    "missing_value",             # a field that should exist and does not
    "unmapped_part_kind",        # a part-kind no counting rule can produce
    "disputed",                  # two admissible readings disagree — carries `on`
    "illegible_source",          # the source states it; we cannot read it
]

# The two kinds a curator cannot close: they need a schema change here.
PLANNING_CLOSES: frozenset[str] = frozenset({"unmodellable_entity", "unmapped_part_kind"})


class EntityRef(BaseModel):
    """§1.1: `EntityRef { kind, id, tenant }` — a thing in the world.

    `kind` is a **str, not a Literal**, and that is v1.2 being obeyed rather than
    laxity: *"`EntityRef.kind` is a closed vocabulary in the registries, on the
    same terms as `TaskCode`, `SourceClass` and `RoleCode`: adding an entry is
    never a breaking change and never an amendment."* A Literal here would make
    every registry addition a code change on this side and a release the other
    team has to wait for — which is the property registry delegation exists to
    prevent.

    It is therefore NOT the same vocabulary as `GapSubject.kind`, and the two must
    not share a field. `GapSubject.kind` is a closed, locale-keyed discriminator
    over the contract's ref union; this is an open registry of entity kinds. They
    were one field until v1.2 gave us the language to separate them.
    """

    kind: str = ""
    id: str
    # §1.1: `TenantId  str | null    null = tenant-agnostic, i.e. Knowledge-global`.
    # `None` and `""` are different facts and both occur, so the type carries both.
    tenant: str | None = None


class GapSubject(BaseModel):
    """WHAT is missing, addressably — the contract's `EntityRef | SlotRef | ParamRef`.

    A discriminated envelope rather than a bare union, because every consumer wants
    the same two things first (which kind, which name) and a three-arm union buys a
    curator queue nothing but three code paths. The structure the union carries is
    kept in full underneath.

    **`slot` is gone, and its absence is the type doing its job.** v1.2 §1.1 marks
    `SlotRef` RESERVED — *"No producer may emit a slot-shaped `Gap.subject` until
    an amendment defines it"* — so the reservation is enforced here rather than
    trusted to every call site. This repo emitted one (`strategy/generator.py`,
    `gap:post_ground`) at the moment of ratification, which is the argument for
    enforcing it in the type: the contract's own survey said we did not.

    **`scope` and `point` are v1.2's `ParamRef`.** A `param` subject naming only
    `max_span_mm` is not addressable: the first real snapshot publishes the SAME
    parameter twice under different `scope.id`s, so a subject without its scope
    names two different holes at once. `point` is structured for the same reason it
    is structured in the contract — a pre-rendered `"exposure_category=D, hvhz=True"`
    cannot be localised, and Hebrew is not a language this system renders English
    fragments into.
    """

    # The closed, locale-keyed discriminator over the contract's ref union. Every
    # value here needs `gaps.subject.<value>` in BOTH bundles
    # (`tests/web/test_locale_bundles.py` derives the key set from this Literal).
    kind: Literal["entity", "param"]
    id: str
    tenant: str | None = None
    # `EntityRef.kind` when this subject IS an entity — the open registry value,
    # carried as data and never used to build a locale key.
    ref_kind: str = ""
    # `ParamRef.scope` — which product or assembly the parameter belongs to.
    scope: EntityRef | None = None
    # `ParamRef.point` — the cell inside the table, or None for the whole table.
    # A dimension's value is a token, a bool, a number — or, on a `range()`
    # dimension, an `Interval` (§1.3, amendment 007). The interval's own shape
    # belongs to the knowledge layer, so it is accepted here as a mapping rather
    # than imported: `core/` describing a `ParameterTable` type would invert the
    # dependency, and a gap only needs to carry the point, not interpret it.
    point: dict[str, str | bool | int | dict] | None = None

    @model_validator(mode="before")
    @classmethod
    def _accept_the_contracts_own_union(cls, data):
        """Parse the contract's `EntityRef | ParamRef`, not only our envelope.

        §1.2.1 writes `subject` as a bare UNION — `{kind, id, tenant}` for an
        `EntityRef`, `{parameter, scope, point}` for a `ParamRef`. The envelope
        below, with its `entity | param` discriminator, is OURS: a deliberate
        internal choice so a curator queue gets one shape instead of three code
        paths. It is not what crosses.

        That deviation had a cost we did not see until real data arrived. The
        publisher implemented amendment 004 exactly as ratified and every one of
        their 65 gaps still failed to parse here, because their `kind` carries the
        REGISTRY value (`element`, `source_document`) where ours expects the
        discriminator. Our envelope was reading their open vocabulary as our
        closed one — and the symptom looked like their defect.

        So the mapping happens at the door, once. Structural detection rather
        than a `kind` lookup: `parameter` identifies a `ParamRef` and `ref_kind`
        identifies our own shape round-tripping, so neither depends on a registry
        value happening not to collide with a discriminator word.

        A shape matching neither is left to fail. `snapshot.load()` quarantines a
        gap that will not parse, so an unrecognised subject is reported to the
        sender rather than guessed at — and `SlotRef` is still RESERVED, so there
        is no third arm to add.
        """
        if not isinstance(data, dict):
            return data
        if "parameter" in data:                     # ParamRef (§1.1)
            return {
                "kind": "param",
                "id": data["parameter"],
                "tenant": data.get("tenant"),
                "scope": data.get("scope"),
                "point": data.get("point"),
            }
        if "ref_kind" in data or data.get("kind") in ("entity", "param"):
            return data                             # already our envelope
        if data.get("kind") == "slot":
            # §1.1: `SlotRef` is RESERVED — *"No producer may emit a slot-shaped
            # `Gap.subject` until an amendment defines it."*
            #
            # This branch is here because the mapper above nearly destroyed that.
            # Reading an unrecognised `kind` as a registry value turned
            # `{kind: "slot"}` into an entity with `ref_kind: "slot"` — silently
            # admitting the one shape the contract forbids, through the very
            # function written to accept the contract's shapes. The reservation
            # has to be refused explicitly, not left to fall through a default.
            raise ValueError(
                "SlotRef is RESERVED (§1.1): no producer may emit a slot-shaped "
                "Gap.subject until an amendment defines it"
            )
        if "kind" in data and "id" in data:          # EntityRef (§1.1)
            return {
                "kind": "entity",
                "ref_kind": data["kind"],
                "id": data["id"],
                "tenant": data.get("tenant"),
            }
        return data

    def key(self) -> str:
        """A stable identity for this subject, scope and point included.

        This is what makes two gaps about the same hole recognisable as the same
        hole — which is not a nicety: the first real snapshot publishes 16
        `uncovered_condition` gaps and `expand()` independently derives the same
        16 from `table.uncovered`, so without an identity the ingest reports 32
        holes where there are 16. It is also what a derived gap id is built from,
        after the scope-less version collided 16 rows into 8 ids.
        """
        tenant = self.tenant if self.tenant is not None else "~null"
        if self.kind == "entity":
            return f"entity:{self.ref_kind}/{self.id}@{tenant}"
        scope = f"{self.scope.kind}/{self.scope.id}" if self.scope else "~unscoped"
        # `json.dumps(sort_keys=True)` for a mapping value rather than `repr`:
        # a dict's repr follows insertion order, so two identical intervals
        # parsed from differently-ordered JSON would key differently and the
        # same hole would count twice.
        point = (";".join(
            f"{k}=" + (json.dumps(v, sort_keys=True) if isinstance(v, dict) else repr(v))
            for k, v in ((k, self.point[k]) for k in sorted(self.point)))
                 if self.point else "~whole-table")
        return f"param:{self.id}@{scope}#{point}@{tenant}"


class SourceRef(BaseModel):
    """Evidence, where there is any — contract §1.1, verbatim.

    ```text
    SourceRef    { id, belongs_to }        belongs_to = content_hash -> SourceDoc
    ```

    > **BINDING.** `SourceRef` is opaque to Planning in every respect except
    > `belongs_to`. That one field exists because a source ref resolves only on
    > the Discovery surface, which §3.2.2 forbids Planning from calling during a
    > run — so without it an opaque id carries **zero admissibility bits into a
    > pinned snapshot**, and a run cannot tell that three of a definition's five
    > citations are superseded approvals.

    So `belongs_to` is the whole point of the type and the one field this side is
    allowed to read. `id` is opaque and stays opaque: do not parse it, do not
    build one, do not infer a page number from it.

    A run-produced gap usually cites nothing — the evidence for *"no row covers
    this"* is the absence itself — so `cites` is empty far more often on this
    side than on a published gap. That is exactly why this type has to be right
    anyway: the direction it is NOT exercised in is the direction it will first
    be handed real data.
    """

    id: str
    belongs_to: str = ""  # content_hash -> SourceDoc; "" only where none was given


# A `because` param value. `bool` is listed BEFORE `int` deliberately: a domain
# point's `hvhz: true` is a boolean, and under a `str | int` annotation pydantic
# admits it through the `int` arm and stores `1` — after which the sentence renders
# "hvhz=1" in both languages and the type has silently lost a fact it was handed.
#
# The nested value is `PointValue`, which is itself allowed to be a mapping: since
# amendment 007 a condition on a `range()` dimension is an `Interval`, whose own
# `min`/`max` are `Quantity` objects — so a point is two levels deep, not one. A
# one-level annotation admitted the outer dict and rejected the interval inside it.
PointValue = str | bool | int | dict
ParamValue = str | bool | int | dict[str, PointValue]


class Because(BaseModel):
    """§1.2.1: `because { code, params }` — the ONLY rendering mechanism a `Gap`
    has. A `Gap` carries no `text_raw` the way a quoted `DocumentWarning` does, so
    a free-text `message` beside `because` would compete with it and, per the
    contract's own argument for warnings turned the other way round, become what
    implementations actually display while the locale path rots.

    §1.2.1 places no ceiling on what a param VALUE is, and the first real snapshot
    proves why: a condition point is structurally a mapping, and flattening it to a
    string at the boundary is the same loss `Quantity.value_raw` exists to prevent
    one type over. A renderer that meets a mapping formats it from its parts; it
    must never be handed a pre-joined English fragment to interpolate.
    """

    code: str
    params: dict[str, ParamValue] = {}


class Gap(BaseModel):
    id: str
    kind: GapKind
    # WHICH half of a `disputed` subject is in dispute — required by that kind and
    # meaningless on the other seven, so it is optional here and enforced below.
    #
    # Not a detail to flatten away: the contract records that **33.3% of the
    # platform's human-gated facts** carry a note that readers did not agree on
    # the applicability bracket — *the value is certain and the conditions are
    # not*. Dropping `on` turns that into an undifferentiated "somebody
    # disagreed" and discards the half that says where to look.
    on: Literal["value", "conditions"] | None = None
    subject: GapSubject
    because: Because
    cites: list[SourceRef] = []
    would_close: str
    closes_by: Literal["knowledge", "planning"] = "knowledge"
    severity: Literal["warns_line", "informational"] = "warns_line"

    def model_post_init(self, _ctx) -> None:
        # The one invariant worth enforcing in the type: the two schema-change
        # kinds cannot claim a curator can close them. Getting this wrong is not
        # a rendering bug, it is a work item nobody can action.
        if self.kind in PLANNING_CLOSES and self.closes_by != "planning":
            raise ValueError(
                f"a {self.kind} gap closes by a schema change in Planning, "
                f"not by {self.closes_by}"
            )
        # `disputed{ on: value | conditions }` is how the contract writes it: the
        # discriminator is part of the kind, not an optional embellishment.
        if self.kind == "disputed" and self.on is None:
            raise ValueError("a disputed gap must say what is disputed: value | conditions")
        if self.kind != "disputed" and self.on is not None:
            raise ValueError(f"`on` is meaningless on a {self.kind} gap")

File: src/fenceai/web/static/js/gaps.js
// Rendering a `Gap` — what the run knows it does not know.
//
// A gap is not a warning with extra fields. A warning says something about the
// plan on screen; a gap says something about the SYSTEM, and it names the work
// item that would remove it. The contract makes two of its fields binding for
// exactly that reason, and both of them are about the reader:
//
//   * `would_close` — "a gap that only says something is missing sends a curator
//     hunting". So it is never behind a click here: it is on the row.
//   * `closes_by` — `knowledge` means a curator on the other side of the
//     boundary can close it; `planning` means it needs a schema change in this
//     repository and no curator can touch it. So the panel GROUPS by it rather
//     than printing it as a chip: a queue that mixes the two is a queue that
//     shows someone work they cannot do, which is the one property the contract
//     says it must not have.
//
// `severity` is the third axis and the smallest: `warns_line` affects a line in
// the plan (and is therefore also a warning row on the plan), `informational`
// does not. It changes the weight of a row, never its group — where the work
// goes is not a function of how loud it is.
//
// THE HONEST PROBLEM, stated on screen rather than papered over. `would_close`
// is generated ENGLISH prose with no code and no params, in a Hebrew-first
// product where every other user-visible string renders from a locale bundle.
// There is no translation layer for it and inventing one would be worse than
// the hole: a machine translation of a work item, presented as the sentence its
// author wrote, is the same defect the design forbids for a manufacturer's
// warning text ("verbatim and untranslated — never offer a translate
// affordance", frontend design §6). So it renders the way this app already
// renders every other piece of verbatim text written by someone else: quoted,
// `lang="en"` and `dir="ltr"` inside the RTL page, and labelled in the reader's
// own language as what it is — a note written for the knowledge curators. The
// long-term answer is a `close.<code>` registry beside the warning registry,
// which is a registry addition and needs no amendment.
//
// Everything here returns an HTML STRING and owns no DOM, exactly like
// `warnings.js`: the editor injects it under `#gaps`, the BOM tab into its own
// body, and neither reaches into the other's subtree.

import { esc } from "./api.js";
import { t } from "./i18n.js";
import { tagOf } from "./structure-data.js";
import { localizedByCode } from "./warnings.js";

// Grouping order, and it is not alphabetical: what a curator can do comes
// first, because that is the group whose items are actionable by the person
// most likely to be reading. `planning` is real work and is shown, never
// hidden — it is just not THEIR work.
const CLOSERS = ["knowledge", "planning"];

// THE TWO VOCABULARIES A SUBJECT CARRIES, AND WHY ONLY ONE OF THEM IS A KEY.
//
// `subject.kind` is CLOSED (`Literal["entity","param"]`, `core/gaps.py`) and is
// the locale-keyed discriminator: every value needs `gaps.subject.<value>` in
// both bundles, and `tests/web/test_locale_bundles.py` derives that set from the
// Literal so adding an arm without its Hebrew fails there.
//
// `subject.ref_kind` and `scope.kind` are `EntityRef.kind`, which v1.2 makes an
// OPEN registry — *"adding an entry is never a breaking change and never an
// amendment"*. A registry entry can therefore arrive from the other team with no
// release on this side, so `t("gaps.subject." + ref_kind)` would put the literal
// string `gaps.subject.fence_model` on screen the day they add a kind, in both
// languages, with nothing red anywhere. They are DATA here: carried, isolated as
// identifiers, never keys. Keeping the two apart is the whole point of v1.2
// splitting them into two fields.
//
// A CONDITION POINT IS NOT A SENTENCE FRAGMENT. `ParamRef.point` is a mapping
// (`{exposure_category: "D", hvhz: true}`) precisely so a renderer can build the
// phrase from its parts; the contract's own note says a pre-joined
// `"exposure_category=D, hvhz=True"` cannot be localised, and Hebrew is not a
// language this system renders English fragments into. So each dimension is
// looked up as a WORD and each value formatted by its type — a boolean becomes
// `כן`/`yes`, not `true`.

/** The word for a condition dimension, or `null` when this repo has none.
 *
 *  Condition dimensions are themselves a registry the other team extends without
 *  asking (CLAUDE.md: *"registry additions are not amendments"*), so this cannot
 *  be a bare `t()`: an unknown dimension must fall back to its raw name rendered
 *  AS a raw name, never to the literal key. The names it does know are the site
 *  fields this app already has words for (`site.hvhz`, `site.exposure_category`)
 *  — one answer per fact, rather than a second Hebrew for the same dimension.
 */
function dimensionWord(name) {
  const key = "site." + name;
  const word = t(key);
  return word === key ? null : word;
}

/** One value from a condition point, formatted by its type.
 *
 *  A boolean is a fact, not a token: `hvhz: true` reads `כן`, because `true` is
 *  neither Hebrew nor English and a reader deciding whether a row applies to
 *  their site should not have to know which. Strings arrive from published
 *  documents and routinely contain inch marks (`49" to 76"`), so they are
 *  escaped and `<bdi class="sku">`-isolated: an unescaped quote inside an
 *  attribute-adjacent template is an XSS hole, and an un-isolated one reorders
 *  the Hebrew sentence around it.
 */
function pointValueHtml(value) {
  if (typeof value === "boolean")
    return `<span class="gap-point-value">${esc(t(value ? "common.yes" : "common.no"))}</span>`;
  return `<bdi class="${typeof value === "number" ? "num" : "sku"}">${esc(value)}</bdi>`;
}

/** The cell in the table this gap is a hole in — dimension by dimension.
 *
 *  Sorted by dimension name so two renderings of the same point cannot differ
 *  by key order, which is the same reason `GapSubject.key()` sorts.
 */
function pointHtml(point) {
  const names = Object.keys(point || {}).sort();
  if (!names.length) return "";
  const cells = names.map((name) => {
    const word = dimensionWord(name);
    const label = word === null
      ? `<bdi class="sku">${esc(name)}</bdi>`
      : `<span class="gap-point-dim">${esc(word)}</span>`;
    return `<span class="gap-point-cell">${label} = ${pointValueHtml(point[name])}</span>`;
  });
  return `<span class="gap-point">${esc(t("gaps.subject.point"))} ${cells.join(", ")}</span>`;
}

/** WHICH product or assembly the parameter belongs to (`ParamRef.scope`).
 *
 *  Not decoration. The first real published snapshot carries `footing_depth_mm`
 *  under two different scopes, so a subject printed without it names two
 *  different holes at once and a curator closing "the footing_depth_mm gap"
 *  closes the wrong one. `scope.kind` is the open registry value and rides along
 *  as part of the identifier rather than as a key.
 */
function scopeHtml(scope) {
  if (!scope || !scope.id) return "";
  const shown = scope.kind ? `${scope.kind}/${scope.id}` : scope.id;
  return `<span class="gap-scope">${esc(t("gaps.subject.scope"))} `
    + `<bdi class="sku">${esc(shown)}</bdi></span>`;
}

/** What the gap is about, named the way the rest of the app names it.
 *
 *  An `entity` subject is a strategy element id, and the structure report knows
 *  its tag — so a gap about a gate reads "A/G1" rather than
 *  `gate@run1:2000-3000`, the same substitution `warnings.js` makes for element
 *  refs. A `param` ref has no tag and is printed verbatim — it is an identifier,
 *  so it gets `.sku` (which forces LTR) and `<bdi>` — and then says WHERE:
 *  its scope and the condition cell inside that scope's table.
 */
function subjectHtml(subject) {
  if (!subject || !subject.id) return "";
  const entity = subject.kind === "entity";
  const shown = entity ? (tagOf(subject.id) || subject.id) : subject.id;
  // The closed discriminator IS a key — but a stored run written before v1.2
  // retired `slot` still deserializes, and a bare `t()` would print
  // `gaps.subject.slot` at it. Same guard `gapRowHtml` gives `kind`.
  const kindWord = t("gaps.subject." + subject.kind);
  const label = kindWord === "gaps.subject." + subject.kind ? subject.kind : kindWord;
  const parts = [`${esc(label)} <bdi class="sku">${esc(shown)}</bdi>`];
  if (entity && subject.ref_kind)
    parts.push(`<span class="gap-ref-kind">${esc(t("gaps.subject.ref_kind"))} `
      + `<bdi class="sku">${esc(subject.ref_kind)}</bdi></span>`);
  if (!entity) {
    const scope = scopeHtml(subject.scope);
    const point = pointHtml(subject.point);
    if (scope) parts.push(scope);
    if (point) parts.push(point);
  }
  const refKindAttr = subject.ref_kind ? ` data-ref-kind="${esc(subject.ref_kind)}"` : "";
  return `<div class="gap-subject meta"${refKindAttr}>${parts.join(" ")}</div>`;
}

/** Evidence, where there is any.
 *
 *  A `SourceRef` is `{ id, belongs_to }` and the contract is explicit that `id`
 *  is OPAQUE to this side — do not parse it, do not infer a page from it. So the
 *  document hash is what is shown, because `belongs_to` is the only field this
 *  side may read and the only one that means anything to a curator holding the
 *  same snapshot; the opaque id rides along as the thing to quote back.
 *
 *  A run-produced gap usually cites nothing — the evidence for "no row covers
 *  this" is the absence itself — so the empty case is the normal one and must
 *  not render an empty "Evidence:" label.
 *
 *  Each citation is now a `.evidence-link` button carrying the opaque id and
 *  `belongs_to` as data attributes — the shared, documented contract
 *  `js/evidence.js` alone interprets (its module header explains why this is
 *  not a cross-module DOM reach). This module still returns a pure HTML
 *  string and wires no listener itself.
 */
function citesHtml(cites) {
  const refs = (cites || []).filter((c) => c && (c.belongs_to || c.id));
  if (!refs.length) return "";
  return `<div class="gap-cites meta">${esc(t("gaps.cites"))} `
    + refs.map((c) => `<button type="button" class="evidence-link sku" `
        + `data-evidence-id="${esc(c.id)}" data-belongs-to="${esc(c.belongs_to || "")}">`
        + `<bdi>${esc(c.belongs_to || c.id)}</bdi></button>`).join(", ")
    + `</div>`;
}

/** One gap, as a row. Exported because the BOM tab attaches single gaps to the
 *  lines they warn, and a second spelling of a gap row is how the two surfaces
 *  come to disagree about what a gap looks like. */
export function gapRowHtml(gap) {
  const severity = gap.severity === "warns_line" ? "warns_line" : "informational";
  // the same sentence the paired StrategyWarning renders, from the same code and
  // the same params — one localizer, so a gap and its warning cannot drift
  const sentence = localizedByCode("warning", gap.because.code, gap.because.params);
  // A `Gap` has no `message` — `because` is its only rendering mechanism (§1.2.1) —
  // so a code neither bundle carries falls back to the bare code itself, not to an
  // English sentence. That is expected for a kind arriving under a code this repo
  // has never seen, and it must be MARKED as such rather than presented as a
  // localized sentence it is not: `t()` returning its own key is exactly that case.
  const localized = t("warning." + gap.because.code) !== "warning." + gap.because.code;
  const kindWord = t("gaps.kind." + gap.kind);
  // `on` belongs to `disputed` alone: WHICH of the two readings disagree. On any
  // other kind the field is absent, and printing a blank there would suggest the
  // gap is making a claim it is not.
  const on = gap.on ? ` <span class="tag">${esc(t("gaps.on." + gap.on))}</span>` : "";
  return `<div class="gap ${esc(severity)}" data-gap-code="${esc(gap.because.code)}"
       data-gap-kind="${esc(gap.kind)}" data-severity="${esc(severity)}"
       data-closes-by="${esc(gap.closes_by)}">
    <div class="gap-head">
      <span class="tag ${esc(severity)}">${esc(t("gaps.severity." + severity))}</span>
      <span class="gap-kind">${esc(kindWord === "gaps.kind." + gap.kind ? gap.kind : kindWord)}</span>${on}
      <span class="sku">${esc(gap.because.code)}</span>
    </div>
    <div class="gap-what"${localized ? "" : ' lang="en" dir="ltr"'}>${sentence}</div>
    ${subjectHtml(gap.subject)}
    <div class="gap-close">
      <span class="gap-close-label">${esc(t("gaps.would_close"))}</span>
      <q class="verbatim gap-verbatim" lang="en" dir="ltr"><bdi>${esc(gap.would_close || "")}</bdi></q>
      <div class="meta">${esc(t("gaps.would_close_note"))}</div>
    </div>
    ${citesHtml(gap.cites)}
  </div>`;
}

/** The whole surface: every gap this run produced, grouped by who can close it.
 *
 *  Empty string when there are none, so a caller concatenates it
 *  unconditionally — and `{ empty: true }` when the caller wants the panel to
 *  SAY there are none, which is a different claim from saying nothing and is
 *  worth making on a surface whose whole subject is absence.
 */
export function gapsPanelHtml(gaps, { empty = false } = {}) {
  const all = (gaps || []).filter(Boolean);
  if (!all.length && !empty) return "";
  let html = `<div class="panel gaps"><h3>${esc(t("gaps.title"))}</h3>`;
  if (!all.length) return html + `<div class="meta">${esc(t("gaps.none"))}</div></div>`;
  html += `<div class="meta">${esc(t("gaps.hint"))}</div>`;
  for (const closer of CLOSERS) {
    const of = all.filter((g) => g.closes_by === closer);
    if (!of.length) continue;
    // warns_line first inside a group: within one person's work, the items that
    // cost a line come before the ones that cost nothing yet.
    const ordered = [...of.filter((g) => g.severity === "warns_line"),
                     ...of.filter((g) => g.severity !== "warns_line")];
    html += `<div class="gap-group" data-closes-by="${esc(closer)}">
      <h4>${esc(t("gaps.group_" + closer))}
        <span class="num">${esc(String(ordered.length))}</span></h4>
      <div class="meta">${esc(t("gaps.group_" + closer + "_hint"))}</div>`;
    for (const gap of ordered) html += gapRowHtml(gap);
    html += `</div>`;
  }
  return html + `</div>`;
}

/** How many gaps warn a LINE — the number a summary line can say without
 *  implying that an informational note costs the plan something. */
export function warnsLineCount(gaps) {
  return (gaps || []).filter((g) => g && g.severity === "warns_line").length;
}
